src/server/clients_info/client_visibility.rs

```diff
@@ -0,0 +1,284 @@
+    pub(super) fn new(policy: VisibilityPolicy) -> Self {
+        match policy {
+            VisibilityPolicy::All => Self::with_filter(VisibilityFilter::All {
+                just_connected: true,
+            }),
+            VisibilityPolicy::Blacklist => Self::with_filter(VisibilityFilter::Blacklist {
+                list: Default::default(),
+                added: Default::default(),
+                removed: Default::default(),
+            }),
+            VisibilityPolicy::Whitelist => Self::with_filter(VisibilityFilter::Whitelist {
+                list: Default::default(),
+                added: Default::default(),
+                removed: Default::default(),
+            }),
+        }
+    }
+
+    /// Creates a new instance with specific filter.
+    fn with_filter(filter: VisibilityFilter) -> Self {
+        Self {
+            filter,
+            entity_state: Default::default(),
+        }
+    }
+
+    /// Resets the filter state to as it was after [`Self::new`].
+    ///
+    /// `entity_state` remains untouched.
+    pub(super) fn clear(&mut self) {
+        match &mut self.filter {
+            VisibilityFilter::All { just_connected } => *just_connected = true,
+            VisibilityFilter::Blacklist {
+                list,
+                added,
+                removed,
+            }
+            | VisibilityFilter::Whitelist {
+                list,
+                added,
+                removed,
+            } => {
+                list.clear();
+                added.clear();
+                removed.clear();
+            }
+        }
+    }
+
+    /// Marks all entities as not "just added" and clears the removed.
+    ///
+    /// Should be called after each tick.
+    pub(crate) fn update(&mut self) {
+        match &mut self.filter {
+            VisibilityFilter::All { just_connected } => *just_connected = false,
+            VisibilityFilter::Blacklist {
+                list,
+                added,
+                removed,
+            } => {
+                for entity in removed.drain() {
+                    list.remove(&entity);
+                }
+                added.clear();
+            }
+            VisibilityFilter::Whitelist {
+                list,
+                added,
+                removed,
+            } => {
+                for entity in added.drain() {
+                    list.insert(entity, false);
+                }
+                removed.clear();
+            }
+        }
+    }
+
+    /// Removes a despawned entity tracked by this client.
+    pub(super) fn remove_despawned(&mut self, entity: Entity) {
+        match &mut self.filter {
+            VisibilityFilter::All { .. } => (),
+            VisibilityFilter::Blacklist {
+                list,
+                added,
+                removed,
+            } => {
+                if let Some(just_removed) = list.get_mut(&entity) {
+                    *just_removed = true;
```

Comment:
Explained in https://github.com/lifescapegame/bevy_replicon/pull/174#discussion_r1451828399.

Reply:
It's a bug, you are right, sorry for the confusion.
That's because I forgot about tests for despawns. I wrote about this before in Discord. Will add a test to verify.